src/devices/IKeyboard.cpp

```diff
@@ -14,27 +24,150 @@ eHIDType IKeyboard::getType() {
+        .model   = rules.model.c_str(),
+        .layout  = rules.layout.c_str(),
+        .variant = rules.variant.c_str(),
+        .options = rules.options.c_str(),
+    };
+
+    const auto CONTEXT = xkb_context_new(XKB_CONTEXT_NO_FLAGS);
+
+    if (!CONTEXT) {
+        Debug::log(ERR, "setKeymap: CONTEXT null??");
         return;
+    }
+
+    clearManuallyAllocd();
+
+    Debug::log(LOG, "Attempting to create a keymap for layout {} with variant {} (rules: {}, model: {}, options: {})", rules.layout, rules.variant, rules.rules, rules.model,
+               rules.options);
+
+    if (!xkbFilePath.empty()) {
+        auto path = absolutePath(xkbFilePath, g_pConfigManager->configCurrentPath);
+
+        if (FILE* const KEYMAPFILE = fopen(path.c_str(), "r"); !KEYMAPFILE)
+            Debug::log(ERR, "Cannot open input:kb_file= file for reading");
+        else {
+            xkbKeymap = xkb_keymap_new_from_file(CONTEXT, KEYMAPFILE, XKB_KEYMAP_FORMAT_TEXT_V1, XKB_KEYMAP_COMPILE_NO_FLAGS);
+            fclose(KEYMAPFILE);
+        }
+    }
+
+    if (!xkbKeymap)
+        xkbKeymap = xkb_keymap_new_from_names(CONTEXT, &XKBRULES, XKB_KEYMAP_COMPILE_NO_FLAGS);
+
+    if (!xkbKeymap) {
+        g_pConfigManager->addParseError("Invalid keyboard layout passed. ( rules: " + rules.rules + ", model: " + rules.model + ", variant: " + rules.variant +
+                                        ", options: " + rules.options + ", layout: " + rules.layout + " )");
+
+        Debug::log(ERR, "Keyboard layout {} with variant {} (rules: {}, model: {}, options: {}) couldn't have been loaded.", rules.layout, rules.variant, rules.rules, rules.model,
+                   rules.options);
+        memset(&XKBRULES, 0, sizeof(XKBRULES));
+
+        currentRules.rules   = "";
+        currentRules.model   = "";
+        currentRules.variant = "";
+        currentRules.options = "";
+        currentRules.layout  = "us";
+
+        xkbKeymap = xkb_keymap_new_from_names(CONTEXT, &XKBRULES, XKB_KEYMAP_COMPILE_NO_FLAGS);
+    }
+
+    // set internal translation state
+    // demo sunao ni ienai
+    xkbStaticState = xkb_state_new(xkbKeymap);
+
+    updateXKBTranslationState(xkbKeymap);
+
+    const auto NUMLOCKON = g_pConfigManager->getDeviceInt(hlName, "numlock_by_default", "input:numlock_by_default");
 
-    xkb_state_unref(xkbTranslationState);
-    xkbTranslationState = nullptr;
+    if (NUMLOCKON == 1) {
+        // lock numlock
+        const auto IDX = xkb_map_mod_get_index(xkbKeymap, XKB_MOD_NAME_NUM);
+
+        if (IDX != XKB_MOD_INVALID)
+            modifiersState.locked |= (uint32_t)1 << IDX;
+    }
+
+    if (modifiersState.locked != 0)
+        keyboardEvents.modifiers.emit(SModifiersEvent{.locked = modifiersState.locked});
+
+    for (size_t i = 0; i < LEDNAMES.size(); ++i) {
+        ledIndexes.at(i) = xkb_map_led_get_index(xkbKeymap, LEDNAMES.at(i));
+        Debug::log(LOG, "xkb: LED index {} (name {}) got index {}", i, LEDNAMES.at(i), modIndexes.at(i));
```

Comment:
unless valgrind and my ctrl +f skills are missing modIndexes never seems to be assigned to anything resulting in 
```
==343356== Conditional jump or move depends on uninitialised value(s)
==343356==    at 0x76CAF8: std::basic_format_context<std::__format::_Sink_iter<char>, char>::iterator std::__format::__formatter_int<char>::format<unsigned int, std::__format::_Sink_iter<char> >(unsigned int, std::basic_format_context<std::__format::_Sink_iter<char>, char>&) const (format:1098)
==343356==    by 0x76CA12: std::basic_format_context<std::__format::_Sink_iter<char>, char>::iterator std::formatter<unsigned int, char>::format<std::__format::_Sink_iter<char> >(unsigned int, std::basic_format_context<std::__format::_Sink_iter<char>, char>&) const (format:2123)
==343356==    by 0x766090: auto std::__format::_Formatting_scanner<std::__format::_Sink_iter<char>, char>::_M_format_arg(unsigned long)::{lambda(auto:1&)#1}::operator()<unsigned int>(unsigned int&) const (format:4049)
==343356==    by 0x765CDD: decltype(auto) std::basic_format_arg<std::basic_format_context<std::__format::_Sink_iter<char>, char> >::_M_visit<std::__format::_Formatting_scanner<std::__format::_Sink_iter<char>, char>::_M_format_arg(unsigned long)::{lambda(auto:1&)#1}>(std::__format::_Formatting_scanner<std::__format::_Sink_iter<char>, char>::_M_format_arg(unsigned long)::{lambda(auto:1&)#1}&&, std::__format::_Arg_t) (format:3513)
==343356==    by 0x765B2F: decltype(auto) std::visit_format_arg<std::__format::_Formatting_scanner<std::__format::_Sink_iter<char>, char>::_M_format_arg(unsigned long)::{lambda(auto:1&)#1}, std::basic_format_context<std::__format::_Sink_iter<char>, char> >(std::__format::_Formatting_scanner<std::__format::_Sink_iter<char>, char>::_M_format_arg(unsigned long)::{lambda(auto:1&)#1}&&, std::basic_format_arg<std::basic_format_context<std::__format::_Sink_iter<char>, char> >) (format:3567)
==343356==    by 0x7654C0: std::__format::_Formatting_scanner<std::__format::_Sink_iter<char>, char>::_M_format_arg(unsigned long) (format:4038)
==343356==    by 0x77795A: std::__format::_Scanner<char>::_M_on_replacement_field() (format:4001)
==343356==    by 0x763609: std::__format::_Scanner<char>::_M_scan() (format:3949)
==343356==    by 0x762FFF: std::__format::_Sink_iter<char> std::__format::__do_vformat_to<std::__format::_Sink_iter<char>, char, std::basic_format_context<std::__format::_Sink_iter<char>, char> >(std::__format::_Sink_iter<char>, std::basic_string_view<char, std::char_traits<char> >, std::basic_format_args<std::basic_format_context<std::__format::_Sink_iter<char>, char> > const&, std::locale const*) (format:4187)
==343356==    by 0x761963: vformat_to<std::__format::_Sink_iter<char> > (format:4217)
==343356==    by 0x761963: std::vformat[abi:cxx11](std::basic_string_view<char, std::char_traits<char> >, std::basic_format_args<std::basic_format_context<std::__format::_Sink_iter<char>, char> >) (format:4252)
==343356==    by 0x8D424F: void Debug::log<unsigned long&, char const* const&, unsigned int&>(LogLevel, std::basic_format_string<char, std::type_identity<unsigned long&>::type, std::type_identity<char const* const&>::type, std::type_identity<unsigned int&>::type>, unsigned long&, char const* const&, unsigned int&) (src/debug/Log.hpp:66)
==343356==    by 0x8D1E25: IKeyboard::setKeymap(IKeyboard::SStringRuleNames const&) (src/devices/IKeyboard.cpp:123)

```

Reply:
IKeyboard.cpp@127. Can you check in your log whether they are assigned? Maybe not all are?

Reply:
line 123 trying to debug log something line 127 assigns i mean

Reply:
ahhh my bad

Reply:
fixed